Update Broker service configuration dynamically

rdhabalia opened the pull request:
### Motivation

as discussed at #184 : sometime it requires to update Broker-serviceConfiguration dynamically with the help of generic api.

### Modifications

- Annotation which defines service-configuration is dynamic
- znode which contains map of dynamic serviceConfiguration's name and value
- REST/AdminTool api to update specific dynamic-configuration by passing name/value 
- API to register listener which can be notified on config value change

### Result

Broker can update serviceConfiguration dynamically and can execute appropriate action on configuration-value change.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/admin/Brokers.java:
```diff
@@ -79,4 +95,79 @@
+     */
+    private void updateServiecConfiguration(String configName, String configValue) {
+        try {
+            Field field = ServiceConfiguration.class.getDeclaredField(configName);
```
Instead of doing the reflection check on each request, could we doing just once (at startup) and keep a `Set<>` of keys that are dynamically updatable?

merlimat commented on pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java:
```diff
@@ -62,6 +62,7 @@
     private long zooKeeperSessionTimeoutMillis = 30000;
     // Time to wait for broker graceful shutdown. After this time elapses, the
     // process will be killed
+    @FieldContext(dynamic = true)
```
I'd leave this PR to implement the "mechanism" and then we can make the actual variables dynamic in subsequent PRs.

rdhabalia replied:
Yes, but in order to write test-cases and perform testing of this feature I take one of this simple attribute in this PR. It helps to write e2e testcases for this feature.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java:
```diff
@@ -824,4 +841,76 @@ public AuthenticationService getAuthenticationService() {
+
+    /**
+     * Update dynamic-ServiceConfiguration with value present into zk-configuraiton-map and register listeners on
+     * dynamic-ServiceConfiguration field to take appropriate action on change of zk-configuraiton-map.
```
zk-configuration-map

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java:
```diff
@@ -824,4 +841,76 @@ public AuthenticationService getAuthenticationService() {
+        // update brokerShutdownTimeoutMs value on listener notification
+        registerConfigurationListener("brokerShutdownTimeoutMs", new Consumer<String>() {
+            @Override
+            public void accept(String shutdownTime) {
```
can abbreviate into a lambda here

merlimat replied:
Actually, for this example, we shouldn't even need to register for notification, right? 

We should make sure that the value `ServiceConfiguration` instance is updated, and then always read it from there, instead of caching it in a variable.

rdhabalia replied:
>Actually, for this example, we shouldn't even need to register for notification, right?

Actually, we are not updating `ServiceConfiguration` instance's field value but we are calling registered listener for this field and it's registered-listener's responsibility to update the value and take appropriate action. 
Because in case of #181 : Only when `lookupRequestPermits` field-value changes we want to take certain action such as update `Semaphore`'s permits. Now, we have watch on entire [dynamicConfigMap](https://github.com/yahoo/pulsar/pull/186/files/206ba5d2d44275a6e061fffa1a791ab7580b0c8e#diff-483f002e2db69883012f75649177d1edR142) and listener can be triggered on every configKey-value change. So, registered listener can compare existing value and new value, and if it's changed then it can take appropriate action and update `ServiceConfiguration`.

rdhabalia replied:
addressed it on [commit](https://github.com/yahoo/pulsar/pull/186/commits/f4adac9c376b63e1e7e4850cacefde29e8ed8481) can you please take a look of it.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java:
```diff
@@ -824,4 +841,76 @@ public AuthenticationService getAuthenticationService() {
+     * @param listener
+     *            : listener which takes appropriate action on config-value change
+     */
+    public void registerConfigurationListener(String configKey, Consumer<String> listener) {
```
Can we make the listener to be working for string, ints, longs? (Maybe leveraging the `FieldParser`)

rdhabalia replied:
yes, but we can have non-primitive [ParameterizedTypes](https://github.com/yahoo/pulsar/blob/master/pulsar-common/src/main/java/com/yahoo/pulsar/common/util/FieldParser.java#L142) also for which we have to pass `Field` to parse the value. So, made appropriate change.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java:
```diff
@@ -824,4 +841,76 @@ public AuthenticationService getAuthenticationService() {
+                            field.setAccessible(true);
+                            field.set(pulsar().getConfiguration(), FieldParser.value(value,field));
+                            if (log.isDebugEnabled()) {
+                                log.debug("Successfully updated {}/{}", key, value);
```
We should print a info log line, for all the overridden variables

merlimat commented on pulsar-client-tools/src/main/java/com/yahoo/pulsar/admin/cli/CmdBrokers.java:
```diff
@@ -48,9 +48,23 @@ void run() throws Exception {
         super("brokers", admin);
         jcommander.addCommand("list", new List());
         jcommander.addCommand("namespaces", new Namespaces());
+        jcommander.addCommand("updateConfig", new UpdateConfigurationCmd());
```
we've been using `-` as separator in other commands. eg: `update-config`

rdhabalia commented:
@merlimat : addressed all the changes, added one more change:
ConfigRegistered should be triggered only if the [configValue has been changed](https://github.com/rdhabalia/pulsar/blob/f4adac9c376b63e1e7e4850cacefde29e8ed8481/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L926)
any thought?

merlimat commented on pulsar-broker/src/test/java/com/yahoo/pulsar/broker/admin/AdminApiTest.java:
```diff
@@ -377,6 +383,56 @@ public void brokers() throws Exception {
+        getPermitZkNodeMethod.setAccessible(true);
+        getPermitZkNodeMethod.invoke(pulsar.getBrokerService());
+        // verify value is updated
+        assertEquals(pulsar.getConfiguration().getBrokerShutdownTimeoutMs(), shutdownTime);
```
The notification from ZK might be delayed in the background thread. In addition since we're doing "dirty" reads (avoid synchronization), it might take a bit more after the new value is visible on the _getter_ method. We should have a retry loop here in the test to give a bit of time.

rdhabalia replied:
I have added this logic in `testUpdateDynamicConfigurationWithZkWatch`.
This test-case directly triggers registerListener because pulsarService started when znode was not created and can't set the watch if node is not exist. so, in this testcase it explicitly triggers registerListener call.

merlimat replied:
Ok, tough is there also the other test that work with the zk watch and verifies we apply the change?

rdhabalia replied:
yes, [testUpdateDynamicConfigurationWithZkWatch](https://github.com/rdhabalia/pulsar/blob/224004d8221c29ecde0de7bf5d8b92acb6b877c6/pulsar-broker/src/test/java/com/yahoo/pulsar/broker/admin/AdminApiTest.java#L404) first creates znode and any further config change happens by zk-watch. I have added retry loop there to wait for refreshed value.

merlimat commented on pulsar-broker/src/test/java/com/yahoo/pulsar/broker/admin/AdminApiTest.java:
```diff
@@ -377,6 +383,56 @@ public void brokers() throws Exception {
+    public void testUpdateDynamicLocalConfiguration() throws Exception {
+
+        // (1) try to update dynamic field
+        final long shutdownTime = 10;
```
Can we verify that shutdown time is != 10 in the initial config?

merlimat commented on pulsar-client-tools/src/main/java/com/yahoo/pulsar/admin/cli/CmdBrokers.java:
```diff
@@ -48,9 +48,23 @@ void run() throws Exception {
         }
     }
 
+    @Parameters(commandDescription = "Update dynamic-serviceConfiguration of broker")
```
Also, it would be useful to have command to print all the config that are stored in ZK. To know what is being overridden.

rdhabalia replied:
added both commands:
- getUpdatableConfigName
- getAllConfigurations to get key,value for all configs

rdhabalia commented on pulsar-client-admin/src/main/java/com/yahoo/pulsar/client/admin/Brokers.java:
```diff
@@ -67,4 +67,13 @@
+     * @param value
+     * @throws PulsarAdminException
+     */
+    void updateConfiguration(String configName, String configValue) throws PulsarAdminException;
```
so, you mean we should rename the method to `overrideDynamicConfiguration`.?

merlimat replied:
Just renaming. I'm not 100% sure about the name, just wanted to make clear that it doesn't update the config in a broker, rather it writes config in ZK that supersedes whatever all the brokers have locally.

rdhabalia replied:
`updateDynamicConfigurationInZk`, `persistDynamicConfigValue`, `storeDynamicConfigValue` ?

rdhabalia replied:
I have updated the document to call it out clearly and also renamed it as `updateDynamicConfiguration`. Do you think if we rename it to something else.?

rdhabalia commented:
@merlimat I have also added `AdminTools` documentation change in this PR.
As we have PR dependencies #274 -> #181 -> this PR. we can merge it if we are fine with the changes.
